hq/app/schedule/IamAudit.scala

```diff
@@ -1,34 +1,55 @@
 import utils.attempt.FailedAttempt
 
 object IamAudit extends Logging {
 
-  def makeIamNotification(flaggedCreds: Map[AwsAccount, Seq[IAMAlertTargetGroup]]): List[IamNotification] = {
-    flaggedCreds.toList.flatMap { case (awsAccount, targetGroups) =>
+  def makeIamNotification(flaggedCredsToNotify: Map[AwsAccount, Seq[IAMAlertTargetGroup]]): List[IamNotification] = {
+    flaggedCredsToNotify.toList.flatMap { case (awsAccount, targetGroups) =>
       if (targetGroups.isEmpty) {
         logger.info(s"found no IAM user issues for ${awsAccount.name}. No notification required.")
         None
       } else {
         targetGroups.map { tg =>
-          logger.info(s"for ${awsAccount.name}, generating iam notification message for ${tg.users.length} user(s) with outdated keys and missing mfa")
-          tg.users.flatMap { user =>
-            val message = createMessage(tg.users, awsAccount)
-            Some(createNotification(awsAccount, tg.targets :+ Account(awsAccount.accountNumber), message, user.username, createDeadline(user.disableDeadline)))
-          }
+          logger.info(s"for ${awsAccount.name}, generating iam notification message for ${tg.users.length} user(s) with outdated keys and and/or missing mfa")
+          val (usersToReceiveWarningAlerts, usersToReceiveFinalAlerts) = sortAlertsIntoWarningOrFinalAlert(tg.users)
+          createNotifications(warning = true, usersToReceiveWarningAlerts, awsAccount, tg.targets) ++ createNotifications(warning = false, usersToReceiveFinalAlerts, awsAccount, tg.targets)
         }
       }
     }.flatten
   }
 
+  private def sortAlertsIntoWarningOrFinalAlert(users: Seq[VulnerableUser]): (Seq[VulnerableUser], Seq[VulnerableUser]) = {
+    users.foldRight[(Seq[VulnerableUser], Seq[VulnerableUser])]((Seq.empty, Seq.empty)) { case (user, (warningAcc, finalAcc)) =>
+      val deadline = createDeadline(user.disableDeadline)
+      if (isWarningAlert(deadline)) (user +: warningAcc, finalAcc)
+      else if (isFinalAlert(deadline)) (warningAcc, user +: finalAcc)
+      else (warningAcc, finalAcc)
+    }
+  }
+
+  private def createNotifications(warning: Boolean, users: Seq[VulnerableUser], awsAccount: AwsAccount, targets: List[Target]): Seq[IamNotification] = {
+    users.map { user =>
+      val message = if (warning) createWarningMessage(awsAccount, users) else createFinalMessage(awsAccount, users)
+      createNotification(awsAccount, targets :+ Account(awsAccount.accountNumber), message, warningSubject(awsAccount), user.username, createDeadline(user.disableDeadline))
+    }
+  }
+
+  private def createDeadline(date: Option[DateTime]): DateTime = date.getOrElse(DateTime.now.plusDays(iamAlertCadence))
+
+  private def isWarningAlert(deadline: DateTime, today: DateTime = DateTime.now): Boolean = {
+    deadline == today.plusWeeks(1) || deadline == today.plusWeeks(3)
+  }
+  private def isFinalAlert(deadline: DateTime, today: DateTime = DateTime.now): Boolean = deadline == today.plusDays(1)
```

Comment:
I think the intention of this function is to compare if `deadline` and `today.plusDays(1)` have the same calendar date (e.g. `2021-04-01`), but from what I gather the `==` operator calls [equals](http://joda-time.sourceforge.net/apidocs/org/joda/time/base/AbstractInstant.html#equals%28java.lang.Object%29) which according to the docs

> Compares this object with the specified object for equality based on the millisecond instant, chronology and time zone.

So I think in practice this will always return false since it's effectively comparing the full time stamp (e.g. `2021-04-01T19:05:52.182+01:00`).

I wasn't able to run these methods in isolation in the cli so I haven't been able to confirm. Am I way off? I think it'd be worthwhile to have some unit tests for this method and `isWarningAlert` to express the intention with some examples. It's very easy for code that uses the calendar to behave in unexpected ways, I've been burned by this kind of thing in the past!

There's some prior art for tests using DateTime in CredentialsReportTest.scala. If this is an issue, possible solutions for this include calling [withTimeAtStartOfDay](https://stackoverflow.com/questions/13764106/comparing-two-dates-using-joda-time) or using the [DateTimeComparator](https://www.joda.org/joda-time/apidocs/org/joda/time/DateTimeComparator.html#getDateOnlyInstance--)

Reply:
@jorgeazevedo you're spot on! 

Thank you so much for identifying this and another huge thank you for identifying a pattern in the codebase that I can follow and potential Joda DateTime functions that I can use. What a star! Amazing!⭐️😃💯